app/models/miq_provision_virt_workflow.rb

```diff
@@ -1044,19 +1053,35 @@ def create_hash_struct_from_vm_or_template(vm_or_template, options)
                  :allocated_disk_storage => vm_or_template.allocated_disk_storage,
                  :v_total_snapshots      => vm_or_template.v_total_snapshots,
                  :evm_object_class       => :Vm}
-    data_hash[:cloud_tenant] = vm_or_template.cloud_tenant if vm_or_template.respond_to?(:cloud_tenant)
-    hash_struct = MiqHashStruct.new(data_hash)
-    hash_struct.operating_system = MiqHashStruct.new(
-      :product_name => vm_or_template.operating_system.product_name
-    ) if vm_or_template.operating_system
-    hash_struct.ext_management_system = MiqHashStruct.new(
-      :name => vm_or_template.ext_management_system.name
-    ) if vm_or_template.ext_management_system
+    data_hash[:cloud_tenant] = vm_or_template.cloud_tenant if vm_or_template.cloud_tenant_id
+    if vm_or_template.operating_system
+      data_hash[:operating_system] = MiqHashStruct.new(:product_name => vm_or_template.operating_system.product_name)
+    end
     if options[:include_datacenter] == true
-      hash_struct.datacenter_name = vm_or_template.owning_blue_folder.try(:parent_datacenter).try(:name)
+      data_hash[:datacenter_name] = vm_or_template.owning_blue_folder.try(:parent_datacenter).try(:name)
+    end
+
+    # Handle EMS data, either with a cache, or with the relation (assumes it is
+    # preloaded usually)
+    if @_ems_allowed_templates_cache
+      # don't have a key, so don't attempt to add to the data_hash
+      if vm_or_template.ems_id
+        ems_id = vm_or_template.ems_id
+
+        # only fetch the ems if not fetched previously
+        unless @_ems_allowed_templates_cache.key?(vm_or_template.ems_id)
+          @_ems_allowed_templates_cache[ems_id] = ExtManagementSystem.find(ems_id).try(:name)
```

Comment:
Would it make sense to just use AR?

```suggestion
          @_ems_allowed_templates_cache[ems_id] = vm_or_template.ext_management_system
```

Reply:
Since https://github.com/ManageIQ/manageiq/pull/18354 was the end goal (for me), I would say no since we are eventually moving away from having any AR objects instantiated at all.

Probably could have done a `pluck` though.

Reply:
Since we are just fetching a single EMS, going to skip adding a pluck for now, since it is not going to amount to much savings.